### PR TITLE
implement observation over oic/res

### PR DESCRIPTION
**Pull request by jkralik**
## Introduce observation over oic/res resource.

By default, the feature is **disabled** and you can enable it by uncommenting macro `OC_DISCOVERY_RESOURCE_OBSERVABLE` in oc_config.h .

### Default(oic.if.ll) and baseline(oic.if.baseline) interfaces

For default(oic.if.ll), baseline you will get notification when the resource is created (by the developer or CREATE operation) or destroyed(by the developer or DELETE operation). It always sends complete representations.

Example for observe with defaul/links interface(oic.if.ll): [oc_res_links.txt](https://github.com/iotivity/iotivity-lite/files/7523162/oc_res_links.txt)
Example for observe with baseline interface(oic.if.baseline): [oic_res_baseline.txt](https://github.com/iotivity/iotivity-lite/files/7523166/oic_res_baseline.txt)

### Batch interface(oic.if.b)
The first notification contains all resources contents of the device where the observer has access via ACLs. After if any change occurs then only the contents of changed resources are sent. When one update changed more resources than one (eg update device name via /oc/con resource), the contents of all changed resources are sent in one notification.

Example:

- First notification - all contents are sent: [oic_res_batch.txt](https://github.com/iotivity/iotivity-lite/files/7523143/oic_res_batch.txt)
- update device name - /oc/con and oic/d as one notification: 
```json
[
    {
        "href": "ocf://a30aa8b6-c1a1-4a26-4ac7-afd12b114fd0/oic/d",
        "rep": {
            "di": "a30aa8b6-c1a1-4a26-4ac7-afd12b114fd0",
            "dmv": "ocf.res.1.3.0",
            "icv": "ocf.2.0.5",
            "n": "devname",
            "piid": "724cf80f-ad93-4d44-76c5-8602ebc25a36"
        }
    },
    {
        "href": "ocf://a30aa8b6-c1a1-4a26-4ac7-afd12b114fd0/oc/con",
        "rep": {
            "n": "devname"
        }
    }
]
```
- create resource /switches:
```json
[
    {
        "href": "ocf://a30aa8b6-c1a1-4a26-4ac7-afd12b114fd0/switches/1",
        "rep": {
            "value": false
        }
    },
    {
        "href": "ocf://a30aa8b6-c1a1-4a26-4ac7-afd12b114fd0/switches",
        "rep": [
            {
                "eps": [
                    {
                        "ep": "coaps://[fe80::a00:27ff:fe98:e38d]:33490"
                    },
                    {
                        "ep": "coaps://10.112.112.12:39973"
                    },
                    {
                        "ep": "coaps+tcp://[fe80::a00:27ff:fe98:e38d]:53059"
                    },
                    {
                        "ep": "coaps+tcp://10.112.112.12:43787"
                    }
                ],
                "href": "/switches/1",
                "if": [
                    "oic.if.a",
                    "oic.if.baseline"
                ],
                "ins": 3670039971,
                "p": {
                    "bm": 3
                },
                "rel": [
                    "hosts"
                ],
                "rt": [
                    "oic.r.switch.binary"
                ]
            }
        ]
    }
]
```
- delete resource /switches/1:
```json
[
    {
        "href": "ocf://a30aa8b6-c1a1-4a26-4ac7-afd12b114fd0/switches",
        "rep": []
    }
]
```

### CODE REVIEW

File: messaging/coap/observe.c [hunk 1]
```diff
@@ -888,194 +887,275 @@ notify_resource_defaults_observer(oc_resource_t *resource,
...
+    return;
+  }
+  oc_resource_t *discover_resource =
+    oc_core_get_resource_by_index(OCF_RES, resource->device);
+
+  if (discover_resource == resource) {
+    return;
+  }
+
+  coap_observer_t *obs = NULL;
```
Comment (Danielius1922) [inline messaging/coap/observe.c:1052]: Can be moved inside the for-loop, it's not used outside of it.

**Review by ondrejtomcik — APPROVED** [review]

File: port/android/oc_config.h [hunk 2]
```diff
@@ -41,7 +41,10 @@ typedef uint64_t oc_clock_time_t;
 #define OC_BLOCK_WISE
 
 // The maximum size of a response to an OBSERVE request, in bytes.
-#define OC_MAX_OBSERVE_SIZE 512
+//#define OC_MAX_OBSERVE_SIZE 512
```
[the remark this answers is not in the document]
Comment (jkralik) [reply port/android/oc_config.h:44]: OC_MAX_OBSERVE_SIZE is still a valid macro. But if it is not set then it uses limit `OC_MAX_APP_DATA_SIZE` https://github.com/iotivity/iotivity-lite/blob/jkralik/feature/observeDiscoveryResource/messaging/coap/observe.c#L83
The proper value was default is `OC_MAX_APP_DATA_SIZE` but by introducing `OC_MAX_OBSERVE_SIZE` was changed so I fix it.

File: messaging/coap/observe.c [hunk 3]
```diff
@@ -1078,6 +1136,165 @@ leave_notify_observers:;
...
+      }
+      batch_obs = batch_obs->next;
+    }
+    if (found) {
+      obs = obs->next;
+      continue;
+    }
+    batch_observer_t *o = oc_memb_alloc(&batch_observers_memb);
+    if (o == NULL) {
+      // TODO remove all associated events to obs and send batch for all
```
[the remark this answers is not in the document]
Comment (jkralik) [reply messaging/coap/observe.c]: Added check and log error.

File: messaging/coap/observe.c [hunk 4]
```diff
@@ -1078,6 +1136,165 @@ leave_notify_observers:;
...
+    OC_DBG("oc_notify_discovery_batch_observers: Issue GET request to "
+           "discovery resource for %s resource\n\n",
+           oc_string(batch_obs->resource->uri));
+    response_buffer.buffer = buffer;
+    response_buffer.buffer_size = OC_MAX_APP_DATA_SIZE;
+    oc_rep_new(response_buffer.buffer, response_buffer.buffer_size);
+    oc_rep_start_links_array();
+    int size_before = oc_rep_get_encoded_payload_size();
+    batch_observer_t *o = batch_obs->next;
+    // TODO process batch_obs->resource == nil
```
[the remark this answers is not in the document]
Comment (jkralik) [reply messaging/coap/observe.c]: Added check and log warning.

File: messaging/coap/observe.c [hunk 5]
```diff
@@ -298,151 +360,217 @@ coap_remove_observer_by_resource(const oc_resource_t *rsc)
...
 #else  /* OC_TCP */
-      if (obs->obs_counter % COAP_OBSERVE_REFRESH_INTERVAL == 0) {
+          if (obs->obs_counter % COAP_OBSERVE_REFRESH_INTERVAL == 0) {
 #endif /* !OC_TCP */
-        OC_DBG("coap_notify_collections: forcing CON notification to check for "
-               "client liveness");
-        notification->type = COAP_TYPE_CON;
+            OC_DBG("send_notification: forcing CON notification to check "
+                   "for "
+                   "client liveness");
```
[the remark this answers is not in the document]
Comment (jkralik) [reply messaging/coap/observe.c]: fixef

File: messaging/coap/observe.c [hunk 6]
```diff
@@ -616,250 +744,121 @@ coap_remove_observers_on_dos_change(size_t device, bool reset)
...
+coap_notify_observers_internal(oc_resource_t *resource,
+                               oc_response_buffer_t *response_buf,
+                               oc_endpoint_t *endpoint)
 {
   if (!resource) {
-    OC_WRN("coap_notify_observers: no resource passed; returning");
+    OC_WRN("coap_notify_observers_internal: no resource passed; returning");
     return 0;
   }
+  oc_resource_t *discover_resource =
```
[the remark this answers is not in the document]
Comment (jkralik) [reply messaging/coap/observe.c]: moved

File: messaging/coap/observe.c (hunk 1 above)
Comment (jkralik) [reply messaging/coap/observe.c]: moved

File: port/android/oc_config.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (Danielius1922) [reply port/android/oc_config.h:44]: Yes, I'm just asking if this isn't a forgotten debugging code. As far as I can see `oc_config.h` files don't usually have commented out constants, so it looks suspicious.

**Review by Danielius1922 — APPROVED** [review]

**Review by marktrayer — APPROVED** [review]

**Review by WAvdBeek — APPROVED** [review]

**Pull request merged**